Use HttpClient instead of HttpWebRequest

mrward opened the pull request:
Fixes VSTS #601187 - Need to move to HttpClient from HttpWebRequest and use nativehandlers on MacOS

mono-addins change - https://github.com/mono/mono-addins/pull/127

mrward committed:
[Core] Add extension point used when creating http clients

New extension point: /MonoDevelop/Core/HttpMessageHandlerProviders
This is used to return a class that will create a HttpMessageHandler
which will be used when creating HttpClient's with the new
HttpClientProvider class. This extension point will allow platform
specific HttpMessageHandlers to be implemented outside of
MonoDevelop.Core.

mrward committed:
[Core] FileService now uses HttpClient

Downloading the welcome page news xml now uses the HttpClient
instead of HttpWebRequest.

mrward committed:
[Core] Default HttpMessageHandler is now proxy aware.

If the proxy is configured for a url then it will be used by the
default HttpMessageHandlerProvider. This allows requests to use the
proxy but only if the proxy does not require authentication.

mrward committed:
[Core] Add ProxyAuthenticationHandler for HttpClients

Added the ProxyAuthenticationHandler and associated classes from
NuGet - https://github.com/NuGet/NuGet.Client

mrward committed:
[Core] Remove Mono proxy auth workaround for HttpClient

Mono workaround is no longer required with Mono 4.8 and later for
bug: https://bugzilla.xamarin.com/show_bug.cgi?id=19594

mrward committed:
[Core] Prompt for non-proxy credentials when using HttpClient

Added the HttpSourceAuthenticationHandler from NuGet which handles
the HTTP status codes Unauthorized (401) and Forbidden (403) and will
request credentials from the credential provider. This will result
in a credential dialog being displayed if the credentials are not
available.

mrward committed:
[Ide] ImageLoader now uses HttpClient

mrward committed:
[Core] Dispose HttpClient response in FileService

Ensure the HttpResponseMessage is disposed. Also removed an exception
handler that was looking for the not modified (304) status code which
is no longer valid since using the HttpClient an exception is no longer
thrown when this status is returned. Also make sure the content
is not loaded into memory since it can be copied directly to a file
by using HttpCompletionOption.ResponseHeadersRead option which
will return the response when the headers are read without reading
the content.

mrward committed:
[Ide] ImageLoader improve memory use when using HttpClient

When requesting the image from a url the image is no longer loaded
completely into memory in the HttpRequestMessage's content. Now
the HttpCompletionOption.ResponseHeadersRead is used and the content
is then copied directly to a file. Also made sure the response is
always disposed. Previously it was only disposed when a file was
downloaded.

mrward committed:
[Core] Allow some HttpClientHandler properties to be configured

The HttpClientProvider now allows a HttpClientSettings to be used
when creating a HttpClient. This allows the following properties to
be configured on the HttpClientHandler:

AllowAutoRedirect
AutomaticDecompression
PreAuthenticate

The above properties can only be set on the HttpClientHandler so
they need to be set when the HttpClient is created.

mrward committed:
[Core] Use HttpClient with Mono.Addins

Updated reference to mono-addins.
Register HttpClientProvider with Mono.Addins so it will use the
HttpClient created by MonoDevelop.Core with its own custom
HttpMessageHandlers.

mrward committed:
[MacPlatform] Add HttpClient options panel

Added a new options panel that can be used to switch between
NSUrlSession and a managed (mono) implementation when using the
HttpClient. This options panel is available from Other - Network
and has a combo box that can switch between the two http client
handler implementations. NSUrlSession is used by default. Currently
this does not affect when http client handler is used by the IDE.

mrward committed:
[MacPlatform] Add HttpMessageHandlerProvider

Support creating a NSUrlSessionHandler to be used by all web requests
made by the IDE. Also support using the default managed
HttpClientHandler if this is enabled.

mrward committed:
[NuGet] Support using native HttpMessageHandlers

Use HttpMessageHandler from MonoDevelop.Core's HttpClientProvider.
NuGet has custom http source authentication, since a NuGet.Config
file can supply credentials, so the HttpClientProvider supports a
SourceAuthenticationRequired property on the HttpClientSettings to
avoid two http source authentication handlers being created.
Added a custom HttpHandlerResourceV3Provider which will use the
HttpMessageHandler from the HttpClientProvider.

Note that the HttpHandlerResource has a HttpClientHandler property
which the NSUrlSessionHandler cannot supply so a dummy instance of
the HttpClientHandler is used if one cannot be found in the
HttpMessageHandler chain.

mrward committed:
[NuGet] Fix credentials not accepted with package sources

When prompted for credentials to access a password protected package
source the credentials were not being sent to the server. The problem
was that the credentials need to be used with the HttpClientHandler's
Credentials which was not being done. Added a new interface
IHttpCredentialsHandler which needs to be implemented by one of
the HttpMessageHandlers so that credentials can be set by the
HttpSourceAuthenticationHandler when the user enters the username
and password.

Note that this is not working with the NSUrlSessionHandler currently.
The managed default handler is working as expected.

mrward committed:
[NuGet] Fix multiple credential dialogs after cancelling the dialog

When the NSUrlSessionHandler is used and a package source needs
credentials which are not stored in the NuGet.Config file the
credential dialog will open when the package source is used. Clicking
the Cancel button would not keep the dialog closed for the current
NuGet action. The request would be retried and the dialog would be
displayed multiple times. The managed http client handler did not
have this problem. To fix this the original NuGet version of the
HttpSourceAuthenticationHandler is now used by the NuGet addin. This
has extra logic based on the current ActivityCorrelationId which
will prevent the dialog from being displayed multiple times.

mrward committed:
[NuGet] Images now loaded using configured http message handler

Images were being loaded by the HttpClientHandler and not the
configured HttpMessageHandler used by the IDE. Also when checking
package sources the HttpClientHandler. Now the creation of the
HttpClient used to load images and check package sources are
available uses the MonoDevelopHttpHandlerResourceV3Provider which
uses a native http message handler is used if configured.

mrward committed:
[Core] Simplify credential service

Remove the caching logic from the CredentialService. There is already
a caching credential provider so this extra caching is not needed.
The CredentialService is now just a simple Task based API that wraps
calls to the ICredentialProvider which has a non-Task based API.

mrward committed:
[NuGet] Do not prompt for proxy credentials in preferences

When checking the package sources in preferences if a proxy is
required and the credentials are not set then every package source
url would trigger the credentials dialog to be displayed. Now the
ProxyAuthenticationHandler can be made non-interactive to avoid this.
This also prevents the Add Packages dialog from prompting for
credentials when attempting to download images that use http instead
of https. The image downloads now do not prompt which works around
this mono problem - https://github.com/mono/mono/issues/10622

mrward committed:
[NuGet] Fix build error after rebasing

NuGet was updated from 4.3 to 4.7. The API has changed in 4.7 so
the CredentialService on the HttpHandlerResourceV3 is now a Lazy type.

mrward committed:
[Core] Fix null reference exception when using native Mac http handler

The HttpSourceCredentials was not being assigned to the credentials
field in the HttpSourceAuthenticationHandler when the native Mac
http message handler was created.

mrward committed:
[Core] Add workaround for Mono using incorrect uri for proxy

Using the HttpClient to make a non-secure http request through a proxy
that requires basic auth would result in the request failing to
authenticate. The problem is that Mono uses the request uri not the
proxy uri when asking the IDE for credentials. This results in the
ProxyCache not finding the cached credentials. As a workaround the
uri is checked to see if there is an associated proxy uri and if so
another attempt is made to get the credentials.

One example of this happening is in the Add Packages dialog where
the EntityFramework image was not being loaded since its image uri
is http://go.microsoft.com/fwlink/?LinkID=288859 and Mono uses this
uri when requesting credentials. https requests as expected, passing
the proxy uri when getting credentials from the ProxyCache.

https://github.com/mono/mono/issues/10622

mrward committed:
[WebReferences] Support connecting through a proxy

Use the WebRequestHelper to provide proxy authentication support
when adding web references. Previously viewing a web reference
through a proxy would fail. The DiscoveryClientProtocol uses
HttpWebRequest and HttpWebResponse so the HttpClient cannot be used.
Instead a MonoDevelopDiscoveryClientProtocol class which derives
from DiscoveryClientProtocol is used. This new class uses the
WebRequestHelper when getting the web response so proxy auth is
now supported.

mrward committed:
[WebReferences] Fix failure to view services requiring auth to view

With a web service that was protected behind basic auth it would
not be possible to view the service in the Web References dialog.
The problem was that the request was not being recreated when the
WebRequestHelper was re-trying the request with the user supplied
credentials. The MonoDevelopDiscoveryClientProtocol has been changed
to re-create the WebRequest instead of re-using the original request.
Also removed the AskCredentials dialog. The WebRequestHelper provides
its own credentials dialog so the AskCredentials dialpog is not
required.

mrward committed:
[Mac] Handle basic auth protected VSTS NuGet package sources

Attempting to use a VSTS NuGet package source would prompt for
credentials but never be able to connect since the credentials were
not used in the request.

The NSUrlSessionHandler currently only supports NTLM responses.
However some WWW-Authenticate basic auth responses are not passed
to the NSUrlSessionHandlerDelegate's DidReceiveChallenge. Whilst
those that start with 'Basic' in the WWW-Authenticate header value are
passed to DidReceiveChallenge, those that start with something else,
such as Bearer, which VSTS NuGet package sources return, do not get
passed to DidReceiveChallenge. To handle this the response is checked
in the IDE and the Authorization http header is added if a credential
can be found and the request is retried again a single time. This
allows VSTS NuGet package sources

mrward committed:
[Debugger] Add Task api to DebugValueConverter

Added a GetValueAsync method which takes a CancellationToken and
returns a Task<T>. The default implementation calls the synchronous
GetValue method. Currently the async method is used when loading
images with the HttpClient. At some point the synchronous methods
should be deprecated and the debugger should use the async task
methods however some parts of the debugger api do not support tasks.

mrward committed:
[Core.Tests] Add http handler tests

Added http handler tests from NuGet for the proxy and http source
handlers. The tests use Moq and to be able to mock the
ICredentialService in MonoDevelop.Core the InternalsVisibleTo
needs to be set for 'ToDynamicProxyGenAssembly2'.

mrward committed:
[NuGet] Add http source auth handler tests

Added tests from NuGet for handler used in the NuGet addin.

mrward committed:
[NuGet] Fix failing test if NSUrlSessionHandler is being used

Disable use of Xamarin.Mac's NSUrlSessionHandler when unit tests
are being run. Xamarin.Mac is not initialized during the tests so
one of the tests that updates Xamarin.Mac to a more recent version
was failing with the NSUrlSessionHandler was being used.

mrward committed:
[Core] Mark WebRequestHelper methods as obsolete

The HttpClientProvider should be used instead wherever possible.

mrward committed:
[WebReferences] Supress Type or member is obsolete warnings

WebReferences addin needs to use the WebRequestHelper due to the
discovery protocol not supporting HttpClient.

Therzok commented on main/src/addins/MacPlatform/BasicAuthenticationHandler.cs:
```diff
@@ -0,0 +1,112 @@
+			int i = str.Length;
+			byte [] result = new byte [i];
+			for (--i; i >= 0; i--)
+				result [i] = (byte)str [i];
```
A string's character can contain multiple bytes. Do we only support ASCII?

mrward replied:
This was taken from Mono's source code - [BasicClient.cs](https://github.com/mono/mono/blob/master/mcs/class/System/System.Net/BasicClient.cs#L48-L56)

Therzok commented on main/src/addins/MacPlatform/Dialogs/HttpClientOptionsPanel.cs:
```diff
@@ -0,0 +1,61 @@
+		{
+			if (control == null) {
+				widget = new HttpClientOptionsWidget ();
+				control = widget.ToGtkWidget ();
```
`new XwtControl (widget)`

Therzok commented on main/src/addins/MacPlatform/MacHttpMessageHandlerProvider.cs:
```diff
@@ -0,0 +1,66 @@
+		HttpMessageHandler CreateNSUrlSessionHandler (Uri uri, HttpClientSettings settings)
+		{
+			var config = NSUrlSessionConfiguration.DefaultSessionConfiguration;
+			config.RequestCachePolicy = NSUrlRequestCachePolicy.ReloadIgnoringLocalCacheData;
```
`NSURLRequestCachePolicy.ReloadRevalidatingCacheData`?

Therzok commented on main/src/addins/MacPlatform/MacPlatformSettings.cs:
```diff
@@ -0,0 +1,44 @@
+{
+	static class MacPlatformSettings
+	{
+		static readonly string UseNSUrlSessionHandlerProperty = "MonoDevelop.MacIntegration.UseNSUrlSessionHandler";
```
`ConfigurationProperty.Create ("MonoDevelop.MacIntegration.UseNSUrlSessionHandler", true)`?

Therzok commented on main/src/addins/MacPlatform/NSUrlSessionCredentialsHandler.cs:
```diff
@@ -0,0 +1,70 @@
+		{
+			bool retry = false;
+			while (true) {
+				var response = await base.SendAsync (request, cancellationToken);
```
```suggestion
				var response = await base.SendAsync (request, cancellationToken).ConfigureAwait (false);
```

Therzok commented on main/src/addins/MonoDevelop.PackageManagement/MonoDevelop.PackageManagement/MonoDevelopServerWarningLogHandler.cs:
```diff
@@ -0,0 +1,32 @@
+		{
+			var configuration = request.GetOrCreateConfiguration ();
+
+			var response = await base.SendAsync (request, cancellationToken);
```
```suggestion
			var response = await base.SendAsync (request, cancellationToken).ConfigureAwait (false);
```

Therzok commented on main/src/addins/MonoDevelop.PackageManagement/MonoDevelop.PackageManagement/NuGetHttpSourceAuthenticationHandler.cs:
```diff
@@ -0,0 +1,230 @@
+				// store the auth state before sending the request
+				var beforeLockVersion = credentials.Version;
+
+				response = await base.SendAsync (request, cancellationToken);
```
```suggestion
				response = await base.SendAsync (request, cancellationToken).ConfigureAwait (false);
```

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core.Web/HttpSourceAuthenticationHandler.cs:
```diff
@@ -0,0 +1,180 @@
+				// store the auth state before sending the request
+				var beforeLockVersion = credentials.Version;
+
+				response = await base.SendAsync (request, cancellationToken);
```
```suggestion
				response = await base.SendAsync (request, cancellationToken).ConfigureAwait (false);
```

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core.Web/ProxyAuthenticationHandler.cs:
```diff
@@ -0,0 +1,172 @@
+				var cacheVersion = credentialCache.Version;
+
+				try {
+					var response = await base.SendAsync (request, cancellationToken);
```
```suggestion
					var response = await base.SendAsync (request, cancellationToken).ConfigureAwait (false);
```

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core.Web/ProxyAuthenticationHandler.cs:
```diff
@@ -0,0 +1,172 @@
+						return response;
+					}
+
+					if (!await AcquireCredentialsAsync (request.RequestUri, cacheVersion, cancellationToken)) {
```
This seems to be a pattern. Can we have some kind of extension method implementing this?

mrward replied:
Not planning to change this code. This is based on the original NuGet code and I do not want to make it very different from the original. Ideally we would use the same code but I suspect that is unlikely to happen.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core.Web/ProxyCache.cs:
```diff
@@ -30,25 +30,72 @@ class ProxyCache : IProxyCache
+			if (credentials == null)
+				throw new ArgumentNullException (nameof (credentials));
+
+			cache.AddOrUpdate (
```
Please use this overload which takes a factory state argument:
https://github.com/dotnet/corefx/blob/1c673ffb9c74e796a9068ec310aaa25121310a12/src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L1091

Then you can pass the credentials as a parameter and not have to capture in the lambda.

mrward replied:
ConcurrentDictionary.AddOrUpdate Overload does not seem to exist with Mono. I guess this is new for .NET Core.

mrward replied:
Seems to be new in .NET Framework 4.7.2 looking at the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate?view=netframework-4.7.2#System_Collections_Concurrent_ConcurrentDictionary_2_AddOrUpdate__1__0_System_Func__0___0__1__System_Func__0__1___0__1____0_) and switching between 4.7.2 and 4.7.1 -

Therzok replied:
Okay, so, I've recently seen this used in roslyn. Just do:
```
Version = Guid.NewGuid();
cache[proxyAddress] = credentials;
```

https://github.com/dotnet/roslyn/issues/33179

mrward committed:
[MacPlatform] Http client refactor after review

Use XwtControl instead of ToGtkWidget.
Use ConfigurationProperty instead of using PropertyService directly.
Use ConfigureAwait (false) in NSUrlSessionCredentialsHandler

mrward committed:
[Core] Use ConfigureAwait false with http message handlers.

mrward committed:
[NuGet] Use ConfigureAwait false with http message handlers.

slluis commented on main/src/addins/MacPlatform/Dialogs/HttpClientOptionsWidget.UI.cs:
```diff
@@ -0,0 +1,76 @@
+
+			var httpClientHandlerLabel = new Label ();
+			httpClientHandlerLabel.Name = nameof (httpClientHandlerLabel);
+			httpClientHandlerLabel.Text = GettextCatalog.GetString ("HttpClient implementation:");
```
Why do we need to offer this option? Can't we just use NSUrlSession?

mrward replied:
Mainly because NSUrlSession can be very strict in what it supports and can fail. Mono's implementation is offered as a fallback if NSUrlSession does not work. One example of this was using Fiddler as a proxy with basic auth - NSUrlSession would not work with it whilst Mono's HttpClient handled it.

Another example is that it is very strict on Basic auth headers. Had to add a workaround in a custom http message handler to support VSTS NuGet package sources so NSUrlSession could be used.

mrward committed:
Merge branch 'master' into http-client

mrward committed:
[NuGet] Fix failing unit tests

Fixed the ArgumentNullException when trying to access Xamarin.Mac's
NSUrlSessionConfiguration.DefaultSessionConfiguration. The tests now
initialize more of the IDE, such as XWT, so Xamarin.Mac is now
initialized avoiding this exception.

mrward commented:
Updated this with latest master.

Therzok committed:
Use a slightly better variant of ConcurrentDictionary setting
